### DISCUSSION TITLE
How can i transfer my res.data to Uint8Array

### DISCUSSION THREAD

**Discussion by broairport** [jimmywarting/StreamSaver.js#264] [Q&A] [answered]
I use axios to get data from my backend .
I print it out using console.log() ,it's an object.I can download it by using **"new blob([res.data])"**
But when i try to use StreamSaver to download it.i just don't know how can i convert my res.data into Uint8Array.
I've tried to use **"new Uint8Array(res.data)"**  to get Uint8Array but it's empty .
By the way i set responseType as "blob"

**Comment by jimmywarting** [comment, marked as answer]
I would recommend against using axios...
### why? 
- b/c axios uses [XMLHttpRequest](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest) to get data. ([ref](https://github.com/axios/axios/tree/master/lib/adapters))
- xhr don't support streaming.
- axios weighs as much as 6.2 kB when minified
- axios don't work within web workers either...
- when you are using xhr to get data then you download everything into memory. 
  - you might just as well just use [filesaver.js](https://github.com/eligrey/FileSaver.js/) to this end.
- using xhr beats the whole purpose of using streamsaver as it's meant to save you from allocating everything into memory.

only fetch supports streaming...
```js
fetch(url).then(res => {
  res.body.pipeTo(destination)
})
```

**Reply by broairport** [reply]
Thanks!

**Reply by broairport** [reply]
@jimmywarting : Hey jimmy.I slice my file from my backend and send it slice by slice,Is pipeTo func can only be called once for a writestream ?I met this problem when i try to keep calling pipeTo func

**Reply by jimmywarting** [reply]
I would really recommend you to use the backend to save the file rather then using StreamSaver + Service worker to mimic what a server dose to save a file it will work better in safari and you would be able to save the file using a background fetch that isn't tied to the websites - meaning that you can close/refresh the page/tab and still download it. using streamsaver gives the user the false expectation that they can navigate away from the page while in fact it's tied to the page feeding the response from the main thread. streamsaver/filesaver and similar alike is only targeted for developers who generates data on the client side and don't have a backend

---

`pipeTo` can be called multiple times but you would have to use [preventClose](https://streams.spec.whatwg.org/#example-identity-transform-usages)
```js
await stream.pipeTo(destination. { preventClose: true })
// get next chunk and do the same thing over again
// when you got the last chunk then you need to close it or use preventClose: false in the last call
```

otherwise you could take the more manual approach of reading all the chunks from each request and corsponding stream and write to the writable stream yourself without ever using pipeTo
```js
const fileStream = streamSaver.createWriteStream('filename.txt')
const writer = fileStream.getWriter()
while (needsWork) {
  const res = await fetch(url)
  const ab = await res.arrayBuffer()
  await writer.write(new Uint8Array(ab))
}
writer.close()
```

**Reply by broairport** [reply]
First of all,thank you for your reply.
Is what you said **"use the backend to save the file"** refered to frontend send file to the backend and backend(server) save it?

**Reply by jimmywarting** [reply]
I'm suggesting that you shouldn't use axios, fetch, XMLhttpRequest, StreamSaver.js, or FileSaver.js

i'm saying that you should try to navigate to the resource that you want to download. it can be either one of:
- `location.href = 'url'` redirect
- A hidden iframe that opens the `url`
- Submitting a `<form action="url" method="POST" target="_blank">` in case if you need to post (hidden) data
- or a click on a link element `<a href="url" download="filename">`

...as long as you respond with a [content-disposition](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Disposition) attachment header to the response
- You can optionally add support for [accept-ranges](https://developer.mozilla.org/en-US/docs/Web/HTTP/Range_requests) - this allows you to retry the download if something fails and you can continue from where you left off.
- also you can set the `content-type` to something like `application/octet-stream` this will for sure make the browser incompatible of rendering the file on a browser so it will default to saving the file instead.

----

This is basically how StreamSaver works. it sets up a service worker on the clients machine. it opens a url (by navigating to that url) and [responds With](https://developer.mozilla.org/en-US/docs/Web/API/FetchEvent/respondWith) a [Response](https://developer.mozilla.org/en-US/docs/Web/API/Response) constructed from a ReadableStream and a content-disposition attachment header. all so that it can emulate a server (backend) and allowing the browser what it is suppose to be doing.

**Reply by jimmywarting** [reply]
whenever you use ajax do get the data then you prevent the browser from doing what it is supposed to do, then content-disposition attachment header has no value/meaning

**Reply by broairport** [reply]
Actually,I am doing something with "Range".Why i choose to use fetch is to retry the download from where it left off if something fails.Beacuse i used to think that browser can't retry from where it left off if i use some DOM element to navigate the resource ,in other words, i should implement it munually.And now i understand what you're saying
Thank you very much!jimmy.

**Reply by broairport** [reply]
What i did is called 喧宾夺主 which means that the wrangling guest robs the place of the host .lol
